system/Database/Query.php

```diff
@@ -297,15 +297,27 @@ public function getErrorMessage(): string
 
 	//--------------------------------------------------------------------
 
+	/**
+	 * Determines if the statement is a write-type query or not.
+	 *
+	 * @param string $sql
+	 * @return boolean
+	 */
+	public static function sqlIsWriteType($sql): bool
+	{
+		return (bool) preg_match('/^\s*"?(SET|INSERT|UPDATE|DELETE|REPLACE|CREATE|DROP|TRUNCATE|LOAD|COPY|ALTER|RENAME|GRANT|REVOKE|LOCK|UNLOCK|REINDEX|OPTIMIZE|MERGE)\s/i', $sql);
+	}
```

Comment:
I think this static function is not needed. Because there is `isWriteType()`.

Reply:
Please review my [comment above](https://github.com/codeigniter4/CodeIgniter4/pull/4176#issuecomment-775565181). The current CI4 code already has numerous different places where it needs to check whether an SQL expression reflects a writable type -- not always in a query. I think the check for writable query needs to be static so it can be invoked elsewhere.

Reply:
Generally speaking, *static* is not good. It might imply bad design.

How about BaseConnection have `isWriteType()`?

Reply:
Thanks for your helpful commentary, @kenjis 

> Generally speaking, _static_ is not good. It might imply bad design.

I'm not sure I agree that static functions are "not good" or "imply bad design." I've never seen any scholarly articles discouraging the use of static functions in a class and would be interested in reading more if you'd like to suggest a link. My reason for using a static function was because I searched the code and found existing isWriteType functions in both QueryInterface and ConnectionInterface classes. I would draw your attention to [my post above](https://github.com/codeigniter4/CodeIgniter4/pull/4176#issuecomment-775565181) where I point out that the current CI4 code currently defines 3 nearly identical isWriteType functions, two in the SQLite3 and SQLSRV connection classes, and once in the base Query class.

> How about `BaseConnection` has `isWriteType()`?

Your suggestion to define this function in the Connection classes seems reasonable, but these classes have no SQL property of their own, so this function would need to receive an `$sql` parameter to check. The reason I thought it should go in the Query class is because queries have SQL associated with them. The QueryInterface class declares this function and it is defined to operate on `$this->originalQueryString`. On the other hand, I just searched and apparently that function is *never* called anywhere in the entire system folder. It is, however, called in [$this->originalQueryString](https://github.com/codeigniter4/CodeIgniter4/blob/develop/tests/system/Database/BaseQueryTest.php#L190). I don't feel like I should necessarily be changing the QueryInterface class, which requires an isWriteType function. If we define the isWriteType function as static in the Connection classes, we could call it from the Query classes without having to instantiate the other class.

I'm going to have to look at this tomorrow. It's late here. In the meantime, I'll need to figure out how to update/merge/rebase my bugs/issue-4069 branch with the latest source from CI4. I had added CI4 to my repo as a remote as a basis for PR #4307 so things are getting pretty complicated in my fork. I expect I'll be confused about the correct git approach.

Reply:
@sneakyimp I appreciate your sincere attitude.

I've also never seen any scholarly articles, but I saw blog articles. You can see them by googling "static method harmful".

The main reason is it breaks OOP princeples.
I just googled and found one: 
https://medium.com/@karmanno/why-are-static-embers-and-methods-bad-for-object-oriented-design-b4d0c141512c

I understand that you say static methods with no side-effects is safe.
I don't insist never use static method.

I think where the method should be.

Reply:
> The main reason is it breaks OOP princeples.
> I just googled and found one:
> https://medium.com/@karmanno/why-are-static-embers-and-methods-bad-for-object-oriented-design-b4d0c141512c

I have tried to understand his points in this article, and I think we satisfy his guidelines for use of static methods:

>Use static methods, when:
>* You need to separate extremely reusable code, which is tightly linked with usage of some tookit or set of classes
>* You need to create a functionality that is tightly coupled with the implementation but isn't linked with any instance. For example data-creation methods or enum converting methods

I would add that the static methods I propose receive as parameter a simple string and return a boolean value. These functions will make no reference at all to objects and will make no assumption about object properties or array keys which may or may not be defined.

I agree that the BaseConnection and DBMS-specific Connection classes would be a good place for this function. I would point out that these various classes do not have any internal SQL property which we might check. Our method would need to receive the `$sql` to be checked as a parameter. Because this function makes no reference to any class properties, it seems an ideal candidate for a static method.